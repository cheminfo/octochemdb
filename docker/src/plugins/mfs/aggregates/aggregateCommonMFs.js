import debugLibrary from '../../../utils/Debug.js';
/**
 * @description Aggregate function for most common mfs in compounds collection
 * @param {*} connection mongo connection
 * @returns {Promise} returns mfsCommon collection
 */
export async function aggregate(connection) {
  const debug = debugLibrary('aggregatemfsCommon');
  // get compounds collection and progress
  const collection = await connection.getCollection('compounds');
  // get progress collection mfsCommon
  const progressCompounds = await connection.getProgress('compounds');
  const progress = await connection.getProgress('mfsCommon');
  await connection.setProgress(progress);
  try {
    if (progressCompounds.seq === progress.seq) {
      debug('Aggregation up-to-date');
      return;
    }
    // set progress to aggregating
    progress.state = 'aggregating';
    await connection.setProgress(progress);
    debug(`mfsCommon: Need to aggregate: ${await collection.count()}`);
    // aggregate compounds with with mfs who have at least 5 entries

    let result = await collection.aggregate(
      [
        { $match: { 'data.nbFragments': 1, 'data.charge': 0 } }, // we don't want charges in MF
        {
          $project: {
            _id: 0,
            mf: '$data.mf',
            em: '$data.em',
            unsaturation: '$data.unsaturation',
            atoms: '$data.atoms',
          },
        },
        {
          $group: {
            _id: '$mf',
            em: { $first: '$em' },
            unsaturation: { $first: '$unsaturation' },
            atoms: { $first: '$atoms' },
            count: { $sum: 1 },
          },
        },

        { $match: { count: { $gte: 5 } } }, // only MFs with at least 5 products in pubchem
        {
          $project: {
            _id: '$_id',
            data: {
              em: '$em',
              atoms: '$atoms',
              unsaturation: '$unsaturation',
              count: '$count',
            },
          },
        },
        { $out: 'mfsCommon_tmp' },
      ],
      {
        allowDiskUse: true, // allow aggregation to use disk if necessary
        maxTimeMS: 60 * 60 * 3000, // 3h
      },
    );
    await result.hasNext();
    const temporaryCollection = await connection.getCollection('mfsCommon_tmp');
    //rename temporary collection to mfsCommon
    await temporaryCollection.rename('mfsCommon', { dropTarget: true });

    // set progress to aggregated
    progress.dateEnd = new Date();
    progress.seq = progressCompounds.seq;
    progress.state = 'aggregated';
    await connection.setProgress(progress);

    return result;
  } catch (e) {
    progress.state = 'error';
    await connection.setProgress(progress);
    if (connection) {
      debug(e.message, { collection: 'mfs', connection, stack: e.stack });
    }
  }
}
