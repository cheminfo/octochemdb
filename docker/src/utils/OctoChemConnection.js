import delay from 'delay';
import dotenv from 'dotenv';
import md5 from 'md5';
import { MongoClient } from 'mongodb';

import debugLibrary from './Debug.js';

dotenv.config();
const debug = debugLibrary('OctoChemConnection');

export function OctoChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

OctoChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

OctoChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

OctoChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

OctoChemConnection.prototype.getImportationLog =
  async function getImportationLog(options) {
    const { collectionName, sources, startSequenceID } = options;
    const logsCollection = await this.getImportationLogsCollection();
    const sourcesHash = md5(JSON.stringify(sources));
    const _id = sourcesHash;
    let logs = await logsCollection.find({ _id }).next();
    if (logs === null) {
      logs = {
        _id,
        collectionName,
        sources,
        sourcesHash,
        dateStart: Date.now(),
        dateEnd: Date.now(),
        startSequenceID,
        endSequenceID: 0,
        status: 'updating',
      };
      await logsCollection.insertOne(logs);
    }
    return logs;
  };
OctoChemConnection.prototype.updateImportationLog =
  async function updateImportationLog(logs) {
    const collection = await this.getImportationLogsCollection();
    logs.dateEnd = Date.now();
    await collection.replaceOne({ _id: logs._id }, logs);
  };

OctoChemConnection.prototype.getImportationLogsCollection =
  async function getCollection() {
    return (await this.getDatabase()).collection('importationLogs');
  };

OctoChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

OctoChemConnection.prototype.setProgress = async function setProgress(
  progress,
) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

OctoChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'updating',
      seq: 0,
      dateStart: Date.now(),
      dateEnd: 0,
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

OctoChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

OctoChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
