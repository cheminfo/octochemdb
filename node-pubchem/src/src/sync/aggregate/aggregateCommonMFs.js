'use strict';

const debug = require('debug')('aggregateCommonMFs');

module.exports = async function aggregateCommonMFs(connection) {
  const collection = await connection.getCollection('compounds');
  debug(
    'commonMFs: Need to aggregate',
    await collection.countDocuments(),
    'entries',
  );
  let result = await collection.aggregate(
    [
      //    { $limit: 1e4 },
      { $match: { nbFragments: 1, charge: 0 } }, // we don't want charges in MF
      { $project: { _id: 0, mf: 1, em: 1, unsaturation: 1, atom: 1 } },
      {
        $group: {
          _id: '$mf',
          em: { $first: '$em' },
          unsaturation: { $first: '$unsaturation' },
          atom: { $first: '$atom' },
          total: { $sum: 1 },
        },
      },
      { $match: { total: { $gte: 5 } } }, // only MFs with at least 5 products in pubchem
      { $out: 'commonMFs' },
    ],
    {
      allowDiskUse: true,
      maxTimeMS: 60 * 60 * 1000, // 1h
    },
  );
  await result.hasNext();

  const collectionCommonMFs = await connection.getCollection('commonMFs');

  await collectionCommonMFs.createIndex({ em: 1 });

  return result;
};
