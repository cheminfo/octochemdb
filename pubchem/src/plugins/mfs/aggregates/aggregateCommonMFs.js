import Debug from '../../../utils/Debug.js';

const debug = Debug('aggregatemfsCommon');

export async function aggregate(connection) {
  const collection = await connection.getCollection('compounds');

  const progressCompounds = await connection.getProgress('compounds');
  const progress = await connection.getProgress('mfsCommon');
  if (progressCompounds.seq === progress.seq) {
    debug('Aggregation up-to-date');
    return;
  }

  debug(`'mfsCommon: Need to aggregate: ${await collection.count()}`);
  let result = await collection.aggregate(
    [
      //
      { $match: { 'data.nbFragments': 1, 'data.charge': 0 } }, // we don't want charges in MF
      { $limit: 1e6 },
      {
        $project: {
          _id: 0,
          mf: '$data.mf',
          em: '$data.em',
          unsaturation: '$data.unsaturation',
          atom: '$data.atom',
        },
      },
      {
        $group: {
          _id: '$mf',
          em: { $first: '$em' },
          unsaturation: { $first: '$unsaturation' },
          atom: { $first: '$atom' },
          total: { $sum: 1 },
        },
      },
      { $match: { total: { $gte: 5 } } }, // only MFs with at least 5 products in pubchem
      { $out: 'mfsCommon' },
    ],
    {
      allowDiskUse: true,
      maxTimeMS: 60 * 60 * 1000, // 1h
    },
  );
  await result.hasNext();

  const collectionmfsCommon = await connection.getCollection('mfsCommon');

  await collectionmfsCommon.createIndex({ em: 1 });

  progress.seq = progressCompounds.seq;
  await connection.setProgress(progress);

  return result;
}
