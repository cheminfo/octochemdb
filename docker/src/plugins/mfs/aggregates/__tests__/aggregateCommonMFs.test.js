import { test, expect } from 'vitest';

import { OctoChemConnection } from '../../../../utils/OctoChemConnection.js';
import { aggregate } from '../aggregateCommonMFs';

test(
  'Aggregation mfsCommon',
  async () => {
    const connection = new OctoChemConnection();
    const compoundsCollection = await connection.getCollection('compounds');
    const activesOrNaturalsCollection =
      await connection.getCollection('activesOrNaturals');
    while (true) {
      if (
        (await compoundsCollection.countDocuments()) === 12 &&
        (await activesOrNaturalsCollection.countDocuments()) === 63
      ) {
        break;
      }
    }
    await aggregate(connection);
    const collection = await connection.getCollection('mfsCommon');
    const collectionEntry = await collection
      .find({
        _id: 'C100H110N4',
      })
      .limit(1);
    const result = await collectionEntry.next();
    if (result?._seq) {
      delete result._seq;
    }
    expect(result).toMatchSnapshot();
    await connection.close();
  },
  { timeout: 50000 },
);
