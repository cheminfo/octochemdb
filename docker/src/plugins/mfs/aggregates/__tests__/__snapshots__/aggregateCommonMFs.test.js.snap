// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Aggregation mfsCommon 1`] = `
{
  "_id": "C100H110N4",
  "atom": "{\\"C\\":100,\\"H\\":110,\\"N\\":4}",
  "count": 5,
  "em": 1366.87304956302,
  "unsaturation": 48,
}
`;
